### ISSUE TITLE
Fine-tuned Mistral 7B model always gets reported as torch_dtype Float32 in config.json 

### ISSUE THREAD

**Issue by MB7979** [axolotl-ai-cloud/axolotl#904]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Config.json that reports the correct torch d_type.

### Current behaviour

Hi,

I'm having a little trouble and I'm not sure if it's a bug or just my own ignorance. I'm fine-tuning Mistral 7B on either 2 or 3 A100s on RunPod. Whenever I do a full fine-tune the config.json at the end reports the torch d_type as "float32" even though I have set it to bf16 in my config.yml. However, if I use Deepspeed the config.json has the correct torch d_type "bfloat16" at the end.

I'm not sure if this is having any bearing on the end model. I don't think so? But I'd like to understand what's going on, as other people don't seem to have float32 in their configs. 

Thanks.

### Steps to reproduce

1. Run a full fine-tune on Mistral 7B with multi-GPUs on Runpod, not using Deepspeed.

### Config yaml

```
base_model: mistralai/Mistral-7B-v0.1
model_type: MistralForCausalLM
tokenizer_type: LlamaTokenizer
is_mistral_derived_model: true

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: /workspace/axolotl/datasets/ConversationalShuffledUpdated.jsonl
    type: orcamini


dataset_prepared_path:
val_set_size: 0.015
output_dir: ./out

sequence_len: 4096
sample_packing: false
pad_to_sequence_len: false

wandb_project: Mistral
wandb_entity:
wandb_watch:
wandb_run_id:
wandb_log_model:

gradient_accumulation_steps: 2
micro_batch_size: 2
eval_batch_size: 2
num_epochs: 5
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.000005

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
eval_steps: 50
eval_table_size: 
eval_table_max_new_tokens:
eval_sample_packing: false
save_steps:
save_total_limit: 4
debug:
deepspeed:
weight_decay: 0.0
fsdp:
fsdp_config:
special_tokens:
  bos_token: "<s>"
  eos_token: "</s>"
  unk_token: "<unk>"
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hello, could you please try it on latest to see if problem still occurs?

Mine seems to be fine (with a different lora config).

```
  "torch_dtype": "bfloat16",
```

**Comment by dandm1** [comment]
I've tried it pretty recently (a few days back) and still had the issue. I will try it with the latest pull.

This problem does not occur with lora or qlora training. It is specific to when I do a full fine-tune of Mistral 7B without Deepspeed. With Deepspeed 2 or 3 enabled the config.json of the outputted model reports the dtype correctly as bfloat16.

**Comment by MB7979** [comment]
I’ve tried it pretty recently (a few days back) and still had the same issue. I will try it with the latest pull.

This problem does not occur with lora or QLoRA training. It is specific to when I do a full fine-tune of Mistral 7B without Deepspeed. With Deepspeed 2 or 3 enabled the config.json of the outputted model reports the dtype correctly as bfloat16.

**Comment by NanoCode012** [comment]
Ok, I can reproduce this now with a full finetune. Leaving this config for repo

```yml
base_model: openaccess-ai-collective/tiny-mistral
flash_attention: true
sequence_len: 1024
bf16: true
lora_r: 32
lora_alpha: 64
lora_dropout: 0.05
lora_target_linear: true
val_set_size: 0.1
special_tokens:
  unk_token: <unk>
  bos_token: <s>
  eos_token: </s>
datasets:
  - path: mhenrichsen/alpaca_2k_test
    type: alpaca
num_epochs: 2
micro_batch_size: 2
gradient_accumulation_steps: 1
output_dir: temp_dir
learning_rate: 0.00001
optimizer: adamw_torch
lr_scheduler: cosine
max_steps: 20
save_steps: 10
eval_steps: 10
dataset_processes: 1
```

Could be due to this casting:
https://github.com/OpenAccess-AI-Collective/axolotl/blob/3678a6c41d051ca6376d013c11c948e55b4c8b4f/src/axolotl/utils/models.py#L519-L528

and not casting it back https://github.com/OpenAccess-AI-Collective/axolotl/blob/3678a6c41d051ca6376d013c11c948e55b4c8b4f/src/axolotl/utils/models.py#L549-L558

**Comment by MB7979** [comment]
Yes, that was it. 

Changing to this resolved it: 

```
# LlamaRMSNorm layers are in fp32 after kbit_training or full finetune, so we need to
    # convert them back to fp16/bf16 for flash-attn compatibility.
    if needs_fa2_dtype or (cfg.flash_attention and (cfg.is_llama_derived_model or cfg.is_mistral_derived_model)):
        LOG.info("converting modules to %s for flash attention", cfg.torch_dtype)
        for name, module in model.named_modules():
            if "norm" in name:
                module.to(cfg.torch_dtype)
            if "lm_head" in name or "embed_tokens" in name:
                if hasattr(module, "weight"):
                    module.to(cfg.torch_dtype)
```

Interestingly, I think the loss profile of the training run I did with this was different, so it may have been having an effect on output. Still need to check how it goes in inference. Thanks for your help.

**Comment by NanoCode012** [comment]
I also got a bit different loss, but it was within 0.05.

A side effect of this issue is that weights are stored in fp32, which isn't necessary. I have opened a PR to address this for Mistral. I haven't tried other models.

**Comment by MB7979** [comment]
Yes, loss changes are minimal. Big gains in VRAM efficiency with flash attention and BF16 working correctly though. I’ve been able to move up to a global batch size of 16 on a fft of Mistral 7B compared with 4 previously.

**Comment by NanoCode012** [comment]
That PR was merged. Closing this.
